pkg/kuadranttools/limitador_mutators.go

```diff
@@ -0,0 +1,133 @@
+		}
+
+		update := false
+
+		// Loop through each option
+		for _, opt := range opts {
+			tmpUpdate := opt(desired, existing)
+			update = update || tmpUpdate
+		}
+
+		return update, nil
+	}
+}
+
+func LimitadorAffinityMutator(desired, existing *limitadorv1alpha1.Limitador) bool {
+	update := false
+	if !reflect.DeepEqual(existing.Spec.Affinity, desired.Spec.Affinity) {
+		existing.Spec.Affinity = desired.Spec.Affinity
+		update = true
+	}
+	return update
+}
+
+func LimitadorReplicasMutator(desired, existing *limitadorv1alpha1.Limitador) bool {
+	update := false
+
+	var existingReplicas = 1
+
+	if existing.Spec.Replicas != nil {
+		existingReplicas = *existing.Spec.Replicas
+	}
+
+	var desiredReplicas = 1
+
+	if desired.Spec.Replicas != nil {
+		desiredReplicas = *desired.Spec.Replicas
+	}
+
+	if desiredReplicas != existingReplicas {
+		existing.Spec.Replicas = &desiredReplicas
+		update = true
+	}
+
+	return update
+}
+
+func LimitadorStorageMutator(desired, existing *limitadorv1alpha1.Limitador) bool {
+	update := false
+	if !reflect.DeepEqual(existing.Spec.Storage, desired.Spec.Storage) {
+		existing.Spec.Storage = desired.Spec.Storage
+		update = true
+	}
+	return update
+}
+
+func LimitadorOwnerRefsMutator(desired, existing *limitadorv1alpha1.Limitador) bool {
+	update := false
+	if !reflect.DeepEqual(existing.OwnerReferences, desired.OwnerReferences) {
+		existing.OwnerReferences = desired.OwnerReferences
+		update = true
+	}
+	return update
+}
+
+func LimitadorRateLimitHeadersMutator(desired, existing *limitadorv1alpha1.Limitador) bool {
+	update := false
+	if !reflect.DeepEqual(existing.Spec.RateLimitHeaders, desired.Spec.RateLimitHeaders) {
+		existing.Spec.RateLimitHeaders = desired.Spec.RateLimitHeaders
+		update = true
+	}
+	return update
+}
+
+func LimitadorTelemetryMutator(desired, existing *limitadorv1alpha1.Limitador) bool {
+	update := false
+	if !reflect.DeepEqual(existing.Spec.Telemetry, desired.Spec.Telemetry) {
+		existing.Spec.Telemetry = desired.Spec.Telemetry
+		update = true
+	}
+	return update
+}
+
+func LimitadorPodDisruptionBudgetMutator(desired, existing *limitadorv1alpha1.Limitador) bool {
+	update := false
+	if !reflect.DeepEqual(existing.Spec.PodDisruptionBudget, desired.Spec.PodDisruptionBudget) {
+		existing.Spec.PodDisruptionBudget = desired.Spec.PodDisruptionBudget
+		update = true
+	}
+	return update
+}
+
+func LimitadorResourceRequirementsMutator(desired, existing *limitadorv1alpha1.Limitador) bool {
+	update := false
+	if !reflect.DeepEqual(existing.Spec.ResourceRequirements, desired.Spec.ResourceRequirements) {
+		existing.Spec.ResourceRequirements = desired.Spec.ResourceRequirements
+		update = true
+	}
+	return update
+}
+
+func LimitadorVerbosityMutator(desired, existing *limitadorv1alpha1.Limitador) bool {
+	update := false
+	if !reflect.DeepEqual(existing.Spec.Verbosity, desired.Spec.Verbosity) {
+		existing.Spec.Verbosity = desired.Spec.Verbosity
+		update = true
+	}
+	return update
+}
```

Comment:
Do you think it's possible to DRY these funcs a bit, with something like...

```go
func UpdateFieldIfNotEqual(existing, desired interface{}, fieldName string) bool {
	existingVal := reflect.ValueOf(existing).Elem().FieldByName(fieldName)
	desiredVal := reflect.ValueOf(desired).Elem().FieldByName(fieldName)

	if existingVal.IsValid() && desiredVal.IsValid() {
		if !existingVal.Equal(desiredVal) {
			existingVal.Set(desiredVal)
			return true
		}
	}
	return false
}

```

Since we are updating fields when are not matching the desired state, then we can populate the fields in an array and call it either in a for loop or reducer fn like...

```go
limitadorSpecFieldNames := make(string, 0)
limitadorSpecFieldNames = append(limitadorSpecFieldNames,"Affinity")

....

for _, fieldName := range limitadorSpecFieldNames {
		if UpdateFieldIfNotEqual(existing, desired, fieldName) {
			update = true
		}
	}
```

not sure if we can even use the limitador spec types for the iteration instead of strings and interface{}

Maybe worth to try... maybe not

Reply:
It seems a great idea to avoid DRY. However, I have few little issues with the proposed implementation

* What is the semantics of that `Equal` method? Each field may need different comparing method. 
* The decision call of which fields needs to be "updated if not equal", belongs to the caller, not to the mutator. 
* TBH: I am not very fan of the reflection lib, only when strictly needed.